base/locks.jl

```diff
@@ -136,41 +194,45 @@ unsafe_convert(::Type{Ptr{Void}}, m::Mutex) = m.handle
 
 function _uv_hook_close(x::Mutex)
     h = x.handle
-    x.handle = C_NULL
-    ccall(:uv_mutex_destroy, Void, (Ptr{Void},), h)
-    Libc.free(h)
-    nothing
+    if h != C_NULL
+        x.handle = C_NULL
+        ccall(:uv_mutex_destroy, Void, (Ptr{Void},), h)
+        Libc.free(h)
+        nothing
+    end
 end
 
-function lock!(m::Mutex)
+function lock(m::Mutex)
```

Comment:
ReadWriteLocks.jl is using the wrong lock API and will not be getting the lock guarantees it seems to expect

For the other two, if I can get multithreading to be reasonably stable by v0.5.1, I'll be happy. Until then, the API is officially experimental and unstable: http://docs.julialang.org/en/latest/stdlib/parallel/#multi-threading